transformer_lens/utils.py

```diff
@@ -799,13 +801,241 @@ def get_device():
+    padding tokens in the past positions for left-padded sequences.
+
+    Args:
+        left_attention_mask (torch.Tensor): The left attention mask, indicating which tokens are
+            not padding tokens. Shape: [batch_size, pos]
+
+    Returns:
+        torch.Tensor: The causal attention mask for left padded sequences. Shape: [batch_size, pos, pos]
+    """
+
+    # Initialize a mask with zeros and the same size as the left attention mask
+    # The mask is 3D, with the same number of positions in both the second and third dimensions
+    mask = (
+        einops.repeat(
+            torch.zeros_like(left_attention_mask),
+            "b pos1 -> b pos1 pos2",
+            pos2=left_attention_mask.shape[1],
+        )
+        .bool()
+        .clone()
+    )
+
+    # Compute the number of attended tokens (non-padding tokens) in each sequence
+    num_attended_tokens_list = left_attention_mask.sum(-1).tolist()
+
+    # For each sequence in the batch...
+    for i, num_attended_tokens in enumerate(num_attended_tokens_list):
+        # ...set the lower triangular part of the last num_attended_tokens positions to 1,
+        # preventing each token from attending to padding tokens in the past positions
+        mask[i, -num_attended_tokens:, -num_attended_tokens:] = torch.tril(
+            torch.ones((num_attended_tokens, num_attended_tokens)).bool()
+        )
+
+    return mask
+
+
+def get_nested_attr(obj, attr_str):
+    """
+    Retrieves a nested attribute from an object based on a dot-separated string.
+
+    For example, if `attr_str` is "a.b.c", this function will return `obj.a.b.c`.
+
+    Args:
+        obj (Any): The object from which to retrieve the attribute.
+        attr_str (str): A dot-separated string representing the attribute hierarchy.
+
+    Returns:
+        Any: The value of the nested attribute.
+    """
+    attrs = attr_str.split(".")
+    for attr in attrs:
+        obj = getattr(obj, attr)
+    return obj
+
+
+def set_nested_attr(obj, attr_str, value):
+    """
+    Sets a nested attribute of an object based on a dot-separated string.
+
+    For example, if `attr_str` is "a.b.c", this function will set the value of `obj.a.b.c` to `value`.
+
+    Args:
+        obj (Any): The object on which to set the attribute.
+        attr_str (str): A dot-separated string representing the attribute hierarchy.
+        value (Any): The value to set for the nested attribute.
+    """
+    attrs = attr_str.split(".")
+
+    # Navigate to the deepest object containing the attribute to be set
+    for attr in attrs[:-1]:
+        obj = getattr(obj, attr)
+
+    # Set the nested attribute's value
+    setattr(obj, attrs[-1], value)
+
+
+class LocallyOverridenDefaults:
+    """
+    Context manager that allows temporary overriding of default values within a model.
+    Once the context is exited, the default values are restored.
+
+    WARNING: This context manager must be used for any function/method that directly accesses
```

Comment:
I'm a bit worried about the cases where people forget to add a context manager, but it's hard to enforce it... One possible way to do this is to do something like the following code, but since `padding_side` is a property of `Tokenizer`, how to deal with it is not very straightforward.

@dataclass
class HookedTransformerConfig:
    ...
    _default_prepend_bos: bool = field(default=True, repr=False)
    _inside_context: bool = field(default=False, init=False, repr=False)

    @property
    def default_prepend_bos(self):
        if not self._inside_context:
            raise ValueError("Direct access to default_prepend_bos without context manager is prohibited.")
        return self._default_prepend_bos

    def set_inside_context(self, state: bool):
        self._inside_context = state